### PR TITLE
Add more generic arithmetic, get in sync with scmutils

### PR DESCRIPTION
**Pull request by sritchie**
This PR adds tests and implementations for the following functions in `sicmutils.generic`:

- `cosh`: hyperbolic cosine
- `sinh`: hyperbolic sine
- `tanh`: hyperbolic tangent, ie sinh/cosh
- `sech`: hyperbolic secant, ie 1/cosh
- `csch`: hyperbolic secant, ie 1/sinh
- `acosh`: inverse hyperbolic cosine, ie, `(= x (cosh (acosh x)))`
- `asinh`: inverse hyperbolic sine, ie, `(= x (sinh (asinh x)))`
- `atanh`: inverse hyperbolic tangent, ie, `(= x (tanh (atanh x)))`

These three methods existed in `sicmutils.env`, but not as extensible generics.
Now they're fully extensible:

- `cot`: cotangent, ie 1/tan
- `sec`: secant, ie 1/cos
- `csc`: cosecant, ie 1/sin

These all work with:

- real and complex numbers
- power series (missing a few implementations, operators and matrices are
  missing the same ones for this reason)
- matrices (square matrices return their power series expansions)
- operators (power series expansion of the operator)
- functions (where they create composition)
- symbolic expressions
- Derivatives and dual numbers! The new functions all work with `D`, the
  forward-mode automatic differentiation operator.

### CODE REVIEW

**Review by littleredcomputer — APPROVED** [review]

File: src/sicmutils/calculus/derivative.cljc [hunk 1]
```diff
@@ -311,25 +311,37 @@
...
+               (g/divide x
+                         (g/+ (g/square x)
+                              (g/square y))))
+             (fn [y x]
+               (g/divide (g/negate y)
+                         (g/+ (g/square x)
+                              (g/square y))))
+             :atan2))
+
+(def ^:private sinh (unary-op g/sinh g/cosh))
```
Comment (sritchie) [inline src/sicmutils/calculus/derivative.cljc:341]: these are the only three new entries here.

File: src/sicmutils/calculus/derivative.cljc [hunk 2]
```diff
@@ -311,25 +311,37 @@
 (def ^:private diff-+ (binary-op g/+ (constantly 1) (constantly 1) :plus))
 (def ^:private diff-- (binary-op g/- (constantly 1) (constantly -1) :minus))
 (def ^:private diff-* (binary-op g/* (fn [_ y] y) (fn [x _] x) :times))
-(def ^:private diff-div (binary-op g/divide
-                                   (fn [_ y] (g/divide 1 y))
-                                   (fn [x y] (g/negate (g/divide x (g/square y)))) :divide))
+(def ^:private diff-div
```
Comment (sritchie) [inline src/sicmutils/calculus/derivative.cljc:314]: I did some rearranging, just added three new entries. The defaults can take care of the rest.

File: src/sicmutils/calculus/derivative.cljc [hunk 3]
```diff
@@ -437,20 +449,29 @@
...
 (define-unary-operation g/cos cos)
-(define-unary-operation g/acos acos)
 (define-unary-operation g/tan tan)
+
+(define-unary-operation g/asin asin)
+(define-unary-operation g/acos acos)
 (define-unary-operation g/atan atan)
+(define-binary-operation g/atan atan2)
+
+(define-unary-operation g/sinh sinh)
```
Comment (sritchie) [inline src/sicmutils/calculus/derivative.cljc:466]: all reshuffling except for these three.

File: src/sicmutils/complex.cljc [hunk 4]
```diff
@@ -110,6 +111,24 @@
 (defmethod g/atan [::complex] [^Complex a] (.atan a))
 (defmethod g/magnitude [::complex] [^Complex a] (.abs a))
 
+
+(defmethod g/cosh [::complex] [^Complex a] (.cosh a))
```
Comment (sritchie) [inline src/sicmutils/complex.cljc:115]: on the JVM, the defaults will take care of the rest. On JS, we have explicit definitions for all of the new methods added.

File: src/sicmutils/function.cljc [hunk 5]
```diff
@@ -239,11 +239,14 @@
...
-(make-unary-operation g/asin)
 (make-unary-operation g/cos)
-(make-unary-operation g/acos)
 (make-unary-operation g/tan)
+(make-unary-operation g/asin)
+(make-unary-operation g/acos)
 (make-unary-operation g/atan)
+(make-unary-operation g/sinh)
+(make-unary-operation g/cosh)
+(make-unary-operation g/tanh)
```
Comment (sritchie) [inline src/sicmutils/function.cljc:249]: defaults can handle the missing ones.

File: src/sicmutils/generic.cljc [hunk 6]
```diff
@@ -145,10 +145,49 @@
...
 (def-generic-function cos 1)
 (def-generic-function sin 1)
-(def-generic-function tan 1)
 (def-generic-function asin 1)
 (def-generic-function acos 1)
 (def-generic-function atan [1 2])
+(def-generic-function cosh 1)
+(def-generic-function sinh 1)
+
+;; Trig functions with default implementations provided.
```
Comment (sritchie) [inline src/sicmutils/generic.cljc:154]: our new functions.

File: src/sicmutils/numbers.cljc [hunk 7]
```diff
@@ -55,21 +55,19 @@
...
+(defmethod g/tan [v/numtype] [a] (Math/tan a))
+
+(defmethod g/cosh [v/numtype] [a] (Math/cosh a))
+(defmethod g/sinh [v/numtype] [a] (Math/sinh a))
+(defmethod g/tanh [v/numtype] [a] (Math/tanh a))
+
 (defmethod g/atan [v/numtype] [a] (Math/atan a))
 (defmethod g/atan [v/numtype v/numtype] [a b] (Math/atan2 a b))
 
-(comment
```
Comment (sritchie) [inline src/sicmutils/numbers.cljc:62]: okay, I've started to unravel the mystery here... more later, but it looks like scmutils handles its type tower a bit differently. Keeping these implemented is the right move, imo.

**Pull request merged**